api/src/main/java/org/pmiops/workbench/api/ProfileController.java

```diff
@@ -433,16 +449,51 @@ private User initializeUserIfNeeded() {
   }
 
   @Override
-  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest){
+  public ResponseEntity<Void> invitationKeyVerification(InvitationVerificationRequest invitationVerificationRequest) {
     verifyInvitationKey(invitationVerificationRequest.getInvitationKey());
     return ResponseEntity.status(HttpStatus.NO_CONTENT).build();
   }
 
-  private void verifyInvitationKey(String invitationKey){
-    if(invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
+  private void verifyInvitationKey(String invitationKey) {
+    if (invitationKey == null || invitationKey.equals("") || !invitationKey.equals(cloudStorageService.readInvitationKey())) {
       throw new BadRequestException(
-        "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+          "Missing or incorrect invitationKey (this API is not yet publicly launched)");
+    }
+  }
+
+  /*
+   * This un-authed API method is limited such that we only allow contact email updates before the user has signed in
```

Comment:
One thing we could consider doing to close the remaining gap would be to generate a nonce on account creation (stored on the user table) and require that here (and for the welcome resend).

Reply:
Hmmmm. I don't see any reason to limit the welcome resend. We just wanted to close the following security gap: If a user sets up an account, has NOT yet done the password reset (forced on first sign in), in this time someone could conceivably reset the email, and have the new password sent to that new email, blocking the initial user from the account they were trying to set up. Resending the initial email to the initial entered contact email is not a gap.

Reply:
My comment here is on the updateContactEmail method.

Reply:
@UrsaStutsman and I had some more follow-up discussion offline, but pasting in here for benefit of the group:

> The gap (Greg's PR also mentions it in the description under "Notes"):
> - An anonymous unauthorized user should not be able to change the contact email on an arbitrary account
> - Right now this is possible between signup and first login
>
> How would a nonce address this:
> Specifically, we would generate a random number and store on the User. We'd return this from createAccount and the client would include it on updateContactEmail (and may as well also include it on resendWelcome, agreed there's less of a security risk on that one). Server side we'd verify this number against what we stored. This allows us to verify that the request is coming from the same client session which originally created the account.